README.md
# simplesmtp

## DEPRECATION NOTICE

This module is deprecated for production use. For SMTP servers use [Haraka](https://www.npmjs.org/package/Haraka), for SMTP clients use [smtp-connection](https://www.npmjs.org/package/smtp-connection).

Simplesmtp is a module written for Node v0.6 and slightly updated for Node v0.8. It does not use Node v0.10 streams and probably is going to have a rocky future with Node v0.12. I do not have time to keep it up to date, the thing probably needs a major rewrite for Node v0.12.

Should be fine though for integration testing purposes.

## Info

This is a module to easily create custom SMTP servers and clients - use SMTP as a first class protocol in Node.JS!

[![Build Status](https://secure.travis-ci.org/andris9/simplesmtp.png)](http://travis-ci.org/andris9/simplesmtp)
[![NPM version](https://badge.fury.io/js/simplesmtp.png)](http://badge.fury.io/js/simplesmtp)

## Version warning!

If you are using node v0.6, then the last usable version of **simplesmtp** is v0.2.7

Current version of simplesmtp is fully supported for Node v0.8+

## Support simplesmtp development

[![Donate to author](https://www.paypalobjects.com/en_US/i/btn/btn_donate_SM.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=DB26KWR2BQX5W)

## SMTP Server

## Simple SMTP server

For a simple inbound only, no authentication SMTP server you can use

    simplesmtp.createSimpleServer([options], requestListener).listen(port);

Example

    simplesmtp.createSimpleServer({SMTPBanner:"My Server"}, function(req){
        req.pipe(process.stdout);
        req.accept();
    }).listen(port);

Properties

  * **req.from** - From address
  * **req.to** - an array of To addresses
  * **req.host** - hostname reported by the client
  * **req.remodeAddress** - client IP address

Methods

  * **req.accept** *([id])* - Accept the message with the selected ID
  * **req.reject** *([message])* - Reject the message with the selected message
  * **req.pipe** *(stream)* - Pipe the incoming data to a writable stream

Events

  * **'data'** *(chunk)* - A chunk (Buffer) of the message.
  * **'end'** - The message has been transferred


## Advanced SMTP server

### Usage

Create a new SMTP server instance with

    var smtp = simplesmtp.createServer([options]);

And start listening on selected port

    smtp.listen(25, [function(err){}]);

SMTP options can include the following:

  * **name** - the hostname of the server, will be used for informational messages
  * **debug** - if set to true, print out messages about the connection
  * **timeout** - client timeout in milliseconds, defaults to 60 000 (60 sec.)
  * **secureConnection** - start a server on secure connection
  * **SMTPBanner** - greeting banner that is sent to the client on connection
  * **requireAuthentication** - if set to true, require that the client must authenticate itself
  * **enableAuthentication** - if set to true, client may authenticate itself but don't have to (as opposed to `requireAuthentication` that explicitly requires clients to authenticate themselves)
  * **maxSize** - maximum size of an e-mail in bytes (currently informational only)
  * **credentials** - TLS credentials (`{key:'', cert:'', ca:['']}`) for the server
  * **authMethods** - allowed authentication methods, defaults to `["PLAIN", "LOGIN"]`
  * **disableEHLO** - if set to true, support HELO command only
  * **ignoreTLS** - if set to true, allow client do not use STARTTLS
  * **disableDNSValidation** - if set, do not validate sender domains
  * **disableSTARTTLS** - if set, do not use STARTTLS

### Example

    var simplesmtp = require("simplesmtp"),
        fs = require("fs");

    var smtp = simplesmtp.createServer();
    smtp.listen(25);

    smtp.on("startData", function(connection){
        console.log("Message from:", connection.from);
        console.log("Message to:", connection.to);
        connection.saveStream = fs.createWriteStream("/tmp/message.txt");
    });

    smtp.on("data", function(connection, chunk){
        connection.saveStream.write(chunk);
    });

    smtp.on("dataReady", function(connection, callback){
        connection.saveStream.end();
        console.log("Incoming message saved to /tmp/message.txt");
        callback(null, "ABC1"); // ABC1 is the queue id to be advertised to the client
        // callback(new Error("Rejected as spam!")); // reported back to the client
    });

### Events

  * **startData** *(connection)* - DATA stream is opened by the client (`connection` is an object with `from`, `to`, `host` and `remoteAddress` properties)
  * **data** *(connection, chunk)* - e-mail data chunk is passed from the client
  * **dataReady** *(connection, callback)* - client is finished passing e-mail data, `callback` returns the queue id to the client
  * **authorizeUser** *(connection, username, password, callback)* - will be emitted if `requireAuthentication` option is set to true. `callback` has two parameters *(err, success)* where `success` is Boolean and should be true, if user is authenticated successfully
  * **validateSender** *(connection, email, callback)* - will be emitted if `validateSender` listener is set up
  * **validateRecipient** *(connection, email, callback)* - will be emitted it `validataRecipients` listener is set up
  * **close** *(connection)* - emitted when the connection to client is closed

## SMTP Client

### Usage

SMTP client can be created with `simplesmtp.connect(port[,host][, options])`
where

  * **port** is the port to connect to
  * **host** is the hostname to connect to (defaults to "localhost")
  * **options** is an optional options object (see below)

### Connection options

The following connection options can be used with `simplesmtp.connect`:

  * **secureConnection** - use SSL
  * **name** - the name of the client server
  * **auth** - authentication object `{user:"...", pass:"..."}` or `{XOAuthToken:"base64data"}`
  * **ignoreTLS** - ignore server support for STARTTLS
  * **tls** - optional options object for `tls.connect`, also applies to STARTTLS. For example `rejectUnauthorized` is set to `false` by default. You can override this option by setting `tls: {rejectUnauthorized: true}`
  * **debug** - output client and server messages to console
  * **logFile** - optional filename where communication with remote server has to be logged
  * **instanceId** - unique instance id for debugging (will be output console with the messages)
  * **localAddress** - local interface to bind to for network connections (needs Node.js >= 0.11.3 for working with tls)
  * **greetingTimeout** (defaults to 10000) - Time to wait in ms until greeting message is received from the server
  * **connectionTimeout** (system default if not set) - Time to wait in ms until the socket is opened to the server
  * **socketTimeout** (defaults to 1 hour) - Time of inactivity until the connection is closed
  * **rejectUnathorized** (defaults to false) - if set to true accepts only valid server certificates. You can override this option with the `tls` option, this is just a shorthand
  * **dsn** - An object with methods `success`, `failure` and `delay`. If any of these are set to true, DSN will be used
  * **enableDotEscaping** set to true if you want to escape dots at the begining of each line. Defaults to false.

### Connection events

Once a connection is set up the following events can be listened to:

  * **'idle'** - the connection to the SMTP server has been successfully set up and the client is waiting for an envelope
  * **'message'** - the envelope is passed successfully to the server and a message stream can be started
  * **'ready'** `(success)` - the message was sent
  * **'rcptFailed'** `(addresses)` - not all recipients were accepted (invalid addresses are included as an array)
  * **'error'** `(err, stage)` - An error occurred. The connection is closed and an 'end' event is emitted shortly. Second argument indicates on which SMTP session stage an error occured.
  * **'end'** - connection to the client is closed

### Sending an envelope

When an `'idle'` event is emitted, an envelope object can be sent to the server.
This includes a string `from` and an array of strings `to` property.

Envelope can be sent with `client.useEnvelope(envelope)`

    // run only once as 'idle' is emitted again after message delivery
    client.once("idle", function(){
        client.useEnvelope({
            from: "me@example.com",
            to: ["receiver1@example.com", "receiver2@example.com"]
        });
    });

The `to` part of the envelope includes **all** recipients from `To:`, `Cc:` and `Bcc:` fields.

If setting the envelope up fails, an error is emitted. If only some (not all)
recipients are not accepted, the mail can still be sent but an `rcptFailed`
event is emitted.

    client.on("rcptFailed", function(addresses){
        console.log("The following addresses were rejected: ", addresses);
    });

If the envelope is set up correctly a `'message'` event is emitted.

### Sending a message

When `'message'` event is emitted, it is possible to send mail. To do this
you can pipe directly a message source (for example an .eml file) to the client
or alternatively you can send the message with `client.write` calls (you also
need to call `client.end()` once the message is completed.

If you are piping a stream to the client, do not leave the `'end'` event out,
this is needed to complete the message sequence by the client.

    client.on("message", function(){
        fs.createReadStream("test.eml").pipe(client);
    });

Once the message is delivered a `'ready'` event is emitted. The event has an
parameter which indicates if the message was transmitted( (true) or not (false)
and another which includes the last received data from the server.

    client.on("ready", function(success, response){
        if(success){
            console.log("The message was transmitted successfully with "+response);
        }
    });

### XOAUTH

**simplesmtp** supports [XOAUTH2 and XOAUTH](https://developers.google.com/google-apps/gmail/oauth_protocol) authentication.

#### XOAUTH2


CHANGELOG.md
# CHANGELOG

## v0.3.34 2014-01-14

  * Bumped version to 0.3.34
  * Fixed a bug with ES6 strict mode (can't set properties to a `false` value)

## v0.3.33 2014-09-04

  * Bumped version to 0.3.33
  * Added deprecation notice

## v0.3.32 2014-05-30

  * Bumped version to 0.3.32
  * ignore close if end was already called [004ebaee]

## v0.3.30 2014-05-13

  * Bumped version to 0.3.30
  * Added .npmignore [a7344b49]

## v0.3.29 2014-05-07

  * Bumped version to 0.3.29
  * Changed formatting rules, use single quotes instead of double quotes [92b581c8]
  * rollback NOOP usage [e47e24bb]

## v0.3.28 2014-05-03

  * Bumped version to 0.3.28
  * handle errors with NOOP [deb18352]

## v0.3.27 2014-04-23

  * Bumped version to 0.3.27
  * get tests running in node 0.8, 0.10, 0.11 [9b3f9043..833388d5]

## v0.3.26 2014-04-23

  * Bumped version to 0.3.26
  * Server: Added support for XOAUTH2 authentication [87b6ed66]
  * Client: Use interval NOOPing to keep the connection up [184d8623]
  * Client: do not throw if recipients are note set [785a2b09]

## v0.3.25 2014-04-16

  * Bumped version to 0.3.25
  * disabled server test for max incoming connections [476f8cf5]
  * Added socketTimeout option [b83a4838]
  * fix invalid tests [cf22d390]

## v0.3.24 2014-03-31

  * Bumped version to 0.3.24
  * Added test for empty MAIL FROM [7f17174d]
  * Allow null return sender in mail command (coxeh) [08bc6a6f]
  * incorrect mail format fix (siterra) [d42d364e]
  * support for `form` and `to` structure: {address:"...",name:"..."} siterra) [2b054740]
  * Improved auth supports detection (finian) [863dc019]
  * Fixed a Buffer building bug (finian) [6dc9a4e2]

## v0.3.23 2014-03-10

  * Bumped version to 0.3.23
  * removed pipelining [4f0a382f]
  * Rename disableDotEscaping to enableDotEscaping [5534bd85]
  * Ignore OAuth2 errors from destroyed connections (SLaks) [e8ff3356]

## v0.3.22 2014-02-16

  * Bumped version to 0.3.22
  * Emit error on unexpected close [111da167]
  * Allowed persistence of custom properties when resetting envelope state. (garbetjie) [b49b7ead]

## v0.3.21 2014-02-16

  * Bumped version to 0.3.21
  * Ignore OAuth errors from destroyed connections (SLaks) [d50a7571]

## v0.3.20 2014-01-28

  * Bumped version to 0.3.20
  * Re-emit 'drain' from tcp socket [5bfb1fcc]

## v0.3.19 2014-01-28

  * Bumped version to 0.3.19
  * Prefer setImmediate over nextTick if available [f53e2d44]
  * Server: Implemented "NOOP" command (codingphil) [707485c0]
  * Server: Allow SIZE with MAIL [3b404028]

## v0.3.18 2014-01-05

  * Bumped version to 0.3.18
  * Added limiting of max client connections (garbetjie) [bcd5c0b3]

## v0.3.17 2014-01-05

  * Bumped version to 0.3.17
  * Do not create a server instance with invalid socket (47d17420)
  * typo (chrisdew) [fe4df83f]
  * Only emit rcptFailed if there actually was an address that was rejected [4c75523f]

## v0.3.16 2013-12-02

  * Bumped version to 0.3.16
  * Expose simplesmtp version number [c2382203]
  * typo in SMTP (chrisdew) [6c39a8d7]
  * Fix typo in README.md (Meekohi) [597a25cb]

## v0.3.15 2013-11-15

  * Bumped version to 0.3.15
  * Fixed bugs in connection timeout implementation (finian) [1a25d5af]

## v0.3.14 2013-11-08

  * Bumped version to 0.3.14
  * fixed: typo causing connection.remoteAddress to be undefined (johnnyleung) 795fe81f
  * improvements to handling stage (mysz) 5a79e6a1
  * fixes TypeError: Cannot use 'in' operator to search for 'dsn' in undefined (mysz) 388d9b82
  * lost saving stage in "DATA" (mysz) de694f67
  * more info on smtp error (mysz) 42a4f964

## v0.3.13 2013-10-29

  * Bumped version to 0.3.13
  * Handling errors which close connection on or before EHLO (mysz) 03345d4d

## v0.3.12 2013-10-29

  * Bumped version to 0.3.12
  * Allow setting maxMessages to pool 5d185708

## v0.3.11 2013-10-22

  * Bumped version to 0.3.11
  * style update 2095d3a9
  * fix tests 17a3632f
  * DSN Support implemented. (irvinzz) d1e8ba29

## v0.3.10 2013-09-09

  * Bumped version to 0.3.10
  * added greetingTimeout, connectionTimeout and rejectUnathorized options to connection pool 8fa55cd3

## v0.3.9 2013-09-09

  * Bumped version to 0.3.9
  * added "use strict" definitions, added new options for client: greetingTimeout, connectionTimeout, rejectUnathorized 51047ae0
  * Do not include localAddress in the options if it is unset 7eb0e8fc

## v0.3.8 2013-08-21

  * Bumped version to 0.3.8
  * short fix for #42, Client parser hangs on certain input (dannycoates) 089f5cd4

## v0.3.7 2013-08-16

  * Bumped version to 0.3.7
  * minor adjustments for better portability with browserify (whiteout-io) 15715498
  * Added raw message to error object (andremetzen) 15715498
  * Passing to error handler the message sent from SMTP server when an error occurred (andremetzen) 15d4cbb4

## v0.3.6 2013-08-06

  * Bumped version to 0.3.6
  * Added changelog
  * Timeout if greeting is not received after connection is established

lib/smtp.js
/**
 * @fileOverview This is the main file for the simplesmtp library to create custom SMTP servers
 * @author <a href="mailto:andris@node.ee">Andris Reinman</a>
 * @version 0.1.2
 */

var RAIServer = require("rai").RAIServer,
    EventEmitter = require('events').EventEmitter,
    oslib = require('os'),
    utillib = require("util"),
    dnslib = require("dns");

module.exports.SMTPServer = SMTPServer;

/**
 * <p>Constructs a SMTP server</p>
 * 
 * <p>Possible options are:</p>
 * 
 * <ul>
 *     <li><b>HOSTNAME</b> - the hostname of the server, will be used for
 *         informational messages</li>
 *     <li><b>debug</b> - if set to true, print out messages about the connection</li>
 *     <li><b>timeout</b> - client timeout in milliseconds, defaults to 60 000</li>
 *     <li><b>SMTPBanner</b> - greeting banner that is sent to the client on connection</li>
 *     <li><b>requireAuthentication</b> - if set to true, require that the client
 *         must authenticate itself</li>
 *     <li><b>validateSender</b> - if set to true, emit <code>'validateSender'</code>
 *         with <code>email</code> and <code>callback</code> when the client
 *         enters <code>MAIL FROM:&lt;address&gt;</code></li>
 *     <li><b>validateRecipients</b> - if set to true, emit <code>'validateRecipient'</code>
 *         with <code>email</code> and <code>callback</code> when the client
 *         enters <code>RCPT TO:&lt;address&gt;</code></li>
 *     <li><b>maxSize</b> - maximum size of an e-mail in bytes</li>
 *     <li><b>credentials</b> - TLS credentials</li>
 *     <li><b>authMethods</b> - allowed authentication methods, defaults to <code>["PLAIN", "LOGIN"]</code></li>
 *     <li><b>disableEHLO</b> - if set, support HELO only</li>
 * </ul>
 * 
 * @constructor
 * @param {Object} [options] Options object
 */
function SMTPServer(options){
    EventEmitter.call(this);
    
    this.options = options || {};
    this.options.HOSTNAME = this.options.HOSTNAME || (oslib.hostname && oslib.hostname()) ||
               (oslib.getHostname && oslib.getHostname()) ||
               "127.0.0.1";

    this.options.authMethods = (this.options.authMethods || ["PLAIN", "LOGIN"]).map(
        function(auth){
            return auth.toUpperCase().trim();
        });
    
    this.options.disableEHLO = !!this.options.disableEHLO;

    this.SMTPServer = new RAIServer({
        timeout: this.options.timeout || 60*1000,
        disconnectOnTimeout: false,
        debug: !!this.options.debug
    });
    
    this.SMTPServer.on("connection", this._createSMTPConnection.bind(this));
}
utillib.inherits(SMTPServer, EventEmitter);

/**
 * Server starts listening on defined port and hostname
 * 
 * @param {Number} port The port number to listen
 * @param {String} [host] The hostname to listen
 * @param {Function} callback The callback function to run when the server is listening
 */
SMTPServer.prototype.listen = function(port, host, callback){
    this.SMTPServer.listen(port, host, callback);
};

/**
 * <p>Closes the server</p>
 * 
 * @param {Function} callback The callback function to run when the server is closed
 */
SMTPServer.prototype.end = function(callback){
    this.SMTPServer.end(callback);
};

/**
 * <p>Creates a new {@link SMTPConnection} object and links the main server with
 * the client socket</p>
 * 
 * @param {Object} client RAISocket object to a client
 */
SMTPServer.prototype._createSMTPConnection = function(client){
    new SMTPConnection(this, client);
};

/**
 * <p>Sets up a handler for the connected client</p>
 * 
 * <p>Restarts the state and sets up event listeners for client actions</p>
 * 
 * @constructor
 * @param {Object} server {@link SMTPServer} instance
 * @param {Object} client RAISocket instance for the client
 */
function SMTPConnection(server, client){
    this.server = server;
    this.client = client;

    this.init();

    if(this.server.options.debug){
        console.log("Connection from", this.client.remoteAddress);
    }
    
    this.client.on("timeout", this._onTimeout.bind(this));
    this.client.on("error", this._onError.bind(this));
    this.client.on("command", this._onCommand.bind(this));
    this.client.on("end", this._onEnd.bind(this));
    
    this.client.on("data", this._onData.bind(this));
    this.client.on("ready", this._onDataReady.bind(this));
    
    // Send the greeting banner
    this.client.send("220 "+this.server.options.HOSTNAME+" "+(this.server.options.SMTPBanner || "ESMTP node.js simplesmtp"));
}

/**
 * <p>Reset the envelope state</p>
 * 
 * <p>If <code>keepAuthData</code> is set to true, then doesn't remove
 * authentication data</p>
 * 
 * @param {Boolean} [keepAuthData=false] If set to true keep authentication data
 */
SMTPConnection.prototype.init = function(keepAuthData){
    this.envelope = {from: "", to:[], date: new Date()};
    
    if(this.hostnamehostNameAppearsAs){
        this.envelope.host = this.hostNameAppearsAs;
    }
    
    if(this.client.remoteAddress){
        this.envelope.remoteAddress = this.client.remoteAddress;
    }
    
    if(!keepAuthData){
        this.authentication = {
            username: false,
            authenticated: false,
            state: "NORMAL"
        };
    }
};

/**
 * <p>Sends a message to the client and closes the connection</p>
 * 
 * @param {String} [message] if set, send it to the client before disconnecting
 */
SMTPConnection.prototype.end = function(message){
    if(message){
        this.client.send(message);
    }
    this.client.end();
};

/**
 * <p>Will be called when the connection to the client is closed</p>
 * 
 * @event
 */
SMTPConnection.prototype._onEnd = function(){
    if(this.server.options.debug){
        console.log("Connection closed to", this.client.remoteAddress);
    }
};

/**
 * <p>Will be called when timeout occurs</p>
 * 
 * @event
 */
SMTPConnection.prototype._onTimeout = function(){
    this.end("421 4.4.2 "+this.server.options.HOSTNAME+" Error: timeout exceeded");
};

/**
 * <p>Will be called when an error occurs</p>
 * 
 * @event
 */
SMTPConnection.prototype._onError = function(){
    this.end("421 4.4.2 "+this.server.options.HOSTNAME+" Error: client error");
};

/**
 * <p>Will be called when a command is received from the client</p>
 * 
 * <p>If there's curently an authentication process going on, route
 * the data to <code>_handleAuthLogin</code>, otherwise act as
 * defined</p>
 * 
 * @event
 * @param {String} command Command
 * @param {Buffer} command Payload related to the command
 */
SMTPConnection.prototype._onCommand = function(command, payload){

    if(this.authentication.state == "AUTHENTICATING"){
        this._handleAuthLogin(command);
        return;
    }
    
    switch(command){
        
        // Should not occur too often 
        case "HELO":
            this._onCommandHELO(payload.toString("utf-8").trim());
            break;
            
        // Lists server capabilities
        case "EHLO":
            if(!this.server.options.disableEHLO){
                this._onCommandEHLO(payload.toString("utf-8").trim());
            }else{
                this.client.send("502 5.5.2 Error: command not recognized");
            }
            break;
            
        // Closes the connection
        case "QUIT":
            this.end("221 2.0.0 Goodbye!");
            break;

        // Resets the current state
        case "RSET":
            this._onCommandRSET();
            break;
        
        // Doesn't work for spam related purposes
        case "VRFY":
            this.end("252 2.1.5 Send some mail, I'll try my best");
            break;
        
        // Initiate an e-mail by defining a sender
        case "MAIL":
            this._onCommandMAIL(payload.toString("utf-8").trim());
            break;
        
        // Add recipients to the e-mail envelope
        case "RCPT":
            this._onCommandRCPT(payload.toString("utf-8").trim());
            break;
        
        // Authenticate if needed
        case "AUTH":
            this._onCommandAUTH(payload);
            break;
        
        // Start accepting binary data stream
        case "DATA":
            this._onCommandDATA();
            break;
        
        // Upgrade connection to secure TLS
        case "STARTTLS":
            this._onCommandSTARTTLS();
            break;
            
        // Display an error on anything else
        default:
            this.client.send("502 5.5.2 Error: command not recognized");
    }  
};

/**
 * <p>Initiate an e-mail by defining a sender.</p>
 * 
 * <p>This doesn't work if authorization is required but the client is
 * not logged in yet.</p>
 * 
 * <p>If <code>validateSender</code> option is set to true, then emits
 * <code>'validateSender'</code> and wait for the callback before moving
 * on</p>
 * 
 * @param {String} mail Address payload in the form of "FROM:&lt;address&gt;"
 */
SMTPConnection.prototype._onCommandMAIL = function(mail){
    var match, email, domain;
    
    if(!this.hostNameAppearsAs){
        return this.client.send("503 5.5.1 Error: send HELO/EHLO first");
    }
    
    if(this.server.options.requireAuthentication && !this.authentication.authenticated){
        return this.client.send("530 5.5.1 Authentication Required");
    }
    
    if(this.envelope.from){
        return this.client.send("503 5.5.1 Error: nested MAIL command");
    }
    
    if(!(match = mail.match(/^from\:\s*<([^@>]+\@([^@>]+))>$/i))){
        return this.client.send("501 5.1.7 Bad sender address syntax");
    }
    
    email = match[1] || "";
    domain = (match[2] || "").toLowerCase();
    
    dnslib.resolveMx(domain, (function(err, addresses){
        if(err || !addresses || !addresses.length){
            return this.client.send("450 4.1.8 <"+email+">: Sender address rejected: Domain not found");
        }
        
        if(this.server.options.validateSender){
            this.server.emit("validateSender", email, (function(err){
                if(err){
                    return this.client.send("550 5.1.1 <"+email+">: Sender address rejected: User unknown in local sender table");
                }
                
                // force domain part to be lowercase
                email = email.substr(0, email.length - domain.length) + domain;
                this.envelope.from = email;
                this.client.send("250 2.1.0 Ok");
        
            }).bind(this));
        }else{
            // force domain part to be lowercase
            email = email.substr(0, email.length - domain.length) + domain;
            this.envelope.from = email;
            this.client.send("250 2.1.0 Ok");
        }
    }).bind(this));  
};

/**
 * <p>Add recipients to the e-mail envelope</p>
 * 
 * <p>This doesn't work if <code>MAIL</code> command is not yet executed</p>
 * 
 * <p>If <code>validateRecipients</code> option is set to true, then emits
 * <code>'validateRecipient'</code> and wait for the callback before moving
 * on</p>
 * 
 * @param {String} mail Address payload in the form of "TO:&lt;address&gt;"
 */
SMTPConnection.prototype._onCommandRCPT = function(mail){
    var match, email, domain;
    
    if(!this.envelope.from){
        return this.client.send("503 5.5.1 Error: need MAIL command");
    }
    
    if(!(match = mail.match(/^to\:\s*<([^@>]+\@([^@>]+))>$/i))){
        return this.client.send("501 5.1.7 Bad recipient address syntax");
    }
    
    email = match[1] || "";
    domain = (match[2] || "").toLowerCase();
    
    dnslib.resolveMx(domain, (function(err, addresses){
        if(err || !addresses || !addresses.length){
            return this.client.send("450 4.1.8 <"+email+">: Recipient address rejected: Domain not found");
        }
        
        if(this.server.options.validateRecipients){
            this.server.emit("validateRecipient", email, (function(err){
                if(err){
                    return this.client.send("550 5.1.1 <"+email+">: Recipient address rejected: User unknown in local recipient table");
                }
                
                // force domain part to be lowercase
                email = email.substr(0, email.length - domain.length) + domain;
                
                // add to recipients list
                if(this.envelope.to.indexOf(email)<0){
                    this.envelope.to.push(email);
                }
                this.client.send("250 2.1.0 Ok");
            }).bind(this));
        }else{
            // force domain part to be lowercase
            email = email.substr(0, email.length - domain.length) + domain;
            
            // add to recipients list
            if(this.envelope.to.indexOf(email)<0){
                this.envelope.to.push(email);
            }
            this.client.send("250 2.1.0 Ok");
        }
    }).bind(this));  
};

/**
 * <p>Switch to data mode and starts waiting for a binary data stream. Emits
 * <code>'startData'</code>.</p>
 * 
 * <p>If <code>RCPT</code> is not yet run, stop</p>
 */
SMTPConnection.prototype._onCommandDATA = function(){
    
    if(!this.envelope.to.length){
        return this.client.send("503 5.5.1 Error: need RCPT command");
    }
    
    this.client.startDataMode();
    this.client.send("354 End data with <CR><LF>.<CR><LF>");
    this.server.emit("startData", this.envelope);
};

/**
 * <p>Resets the current state - e-mail data and authentication info</p>
 */
SMTPConnection.prototype._onCommandRSET = function(){
    this.init();
    this.client.send("250 2.0.0 Ok");
};

/**
 * <p>If the server is in secure connection mode, start the authentication
 * process. Param <code>payload</code> defines the authentication mechanism.</p>
 * 
 * <p>Currently supported - PLAIN and LOGIN. There is no need for more
 * complicated mechanisms (different CRAM versions etc.) since authentication
 * is only done in secure connection mode</p>
 * 
 * @param {Buffer} payload Defines the authentication mechanism
 */
SMTPConnection.prototype._onCommandAUTH = function(payload){
    var method;
    
    if(!this.server.options.requireAuthentication){
        return this.client.send("503 5.5.1 Error: authentication not enabled");
    }
    
    if(!this.client.secureConnection){
        return this.client.send("530 5.7.0 Must issue a STARTTLS command first");
    }
    
    if(this.authentication.authenticated){
        return this.client.send("503 5.7.0 No identity changes permitted");
    }
    
    payload = payload.toString("utf-8").trim().split(" ");
    method = payload.shift().trim().toUpperCase();
    
    if(this.server.options.authMethods.indexOf(method)<0){
        return this.client.send("535 5.7.8 Error: authentication failed: no mechanism available");
    }
    
    switch(method){
        case "PLAIN":
            this._handleAuthPlain(payload);
            break;
        case "LOGIN":
            this._handleAuthLogin();
            break;
    }
};

/**
 * <p>Upgrade the connection to a secure TLS connection</p>
 */
SMTPConnection.prototype._onCommandSTARTTLS = function(){
    if(this.client.secureConnection){
        return this.client.send("554 5.5.1 Error: TLS already active");
    }
    
    this.client.send("220 2.0.0 Ready to start TLS");
    
    this.client.startTLS(this.server.options.credentials, (function(){
        // Connection secured
        // nothing to do here, since it is the client that should 
        // make the next move
    }).bind(this));
};

/**
 * <p>Retrieve hostname from the client. Not very important, since client
 * IP is already known and the client can send fake data</p>
 * 
 * @param {String} host Hostname of the client
 */
SMTPConnection.prototype._onCommandHELO = function(host){
    if(!host){
        return this.client.send("501 Syntax: EHLO hostname");
    }else{
        this.hostNameAppearsAs = host;
        this.envelope.host = host;
    }
    this.client.send("250 "+this.server.options.HOSTNAME+" at your service, ["+
        this.client.remoteAddress+"]");
};

/**
 * <p>Retrieve hostname from the client. Not very important, since client
 * IP is already known and the client can send fake data</p>
 * 
 * <p>Additionally displays server capability list to the client</p>
 * 
 * @param {String} host Hostname of the client 
 */
SMTPConnection.prototype._onCommandEHLO = function(host){
    var response = [this.server.options.HOSTNAME+" at your service, ["+
        this.client.remoteAddress+"]", "8BITMIME", "ENHANCEDSTATUSCODES"];
    
    if(this.server.options.maxSize){
        response.push("SIZE "+this.server.options.maxSize);
    }
    
    if(this.client.secureConnection && this.server.options.requireAuthentication){
        response.push("AUTH "+this.server.options.authMethods.join(" "));
        response.push("AUTH="+this.server.options.authMethods.join(" "));
    }
    
    if(!this.client.secureConnection){
        response.push("STARTTLS");
    }
    
    if(!host){
        return this.client.send("501 Syntax: EHLO hostname");
    }else{
        this.hostNameAppearsAs = host;
        this.envelope.host = host;
    }
    
    this.client.send(response.map(function(feature, i, arr){
        return "250"+(i<arr.length-1?"-":" ")+feature;
    }).join("\r\n"));    
};

/**
 * <p>Detect login information from the payload and initiate authentication
 * by emitting <code>'authorizeUser'</code> and waiting for its callback</p>
 * 
 * @param {Buffer} payload AUTH PLAIN login information
 */
SMTPConnection.prototype._handleAuthPlain = function(payload){
    var userdata = new Buffer(payload.join(" "), "base64"), password;
    userdata = userdata.toString("utf-8").split("\u0000");
    this.authentication.username = userdata[1] || userdata[0] || "";
    password = userdata[2] || "";
    
    this.server.emit("authorizeUser", 
        this.envelope, 
        this.authentication.username, 
        password, 
        (function(err, success){
            if(err || !success){
                this.authentication.authenticated = false;
                this.authentication.username = false;
                this.authentication.state = "NORMAL";
                return this.client.send("535 5.7.8 Error: authentication failed: generic failure");
            }
            this.client.send("235 2.7.0 Authentication successful");
            this.authentication.authenticated = true;
            this.authentication.state = "AUTHENTICATED";
        }).bind(this));
};

/**
 * <p>Sets authorization state to "AUTHENTICATING" and reuqests for the
 * username and password from the client</p>
 * 
 * <p>If username and password are set initiate authentication
 * by emitting <code>'authorizeUser'</code> and waiting for its callback</p>
 * 
 * @param {Buffer} payload AUTH LOGIN login information
 */
SMTPConnection.prototype._handleAuthLogin = function(payload){
    if(this.authentication.state == "NORMAL"){
        this.authentication.state = "AUTHENTICATING";
        this.client.send("334 VXNlcm5hbWU6");
    }else if(this.authentication.state == "AUTHENTICATING"){
        if(this.authentication.username === false){
            this.authentication.username = new Buffer(payload, "base64").toString("utf-8");
            this.client.send("334 UGFzc3dvcmQ6");
        }else{
            this.authentication.state = "VERIFYING";
            this.server.emit("authorizeUser", 
                this.envelope, 
                this.authentication.username, 
                new Buffer(payload, "base64").toString("utf-8"), 
                (function(err, success){
                    if(err || !success){
                        this.authentication.authenticated = false;
                        this.authentication.username = false;
                        this.authentication.state = "NORMAL";
                        return this.client.send("535 5.7.8 Error: authentication failed: generic failure");
                    }
                    this.client.send("235 2.7.0 Authentication successful");
                    this.authentication.authenticated = true;
                    this.authentication.state = "AUTHENTICATED";
                }).bind(this));
        }
        
    }
};

/**
 * <p>Emits the data received from the client with <code>'data'</code>
 * 
 * @event
 * @param {Buffer} chunk Binary data sent by the client on data mode
 */
SMTPConnection.prototype._onData = function(chunk){
    this.server.emit("data", this.envelope, chunk);
};

/**
 * <p>If the data stream ends, emit <code>'dataReady'</code>and wait for
 * the callback.</p>
 * 
 * @event
 */
SMTPConnection.prototype._onDataReady = function(){
    this.server.emit("dataReady", this.envelope, (function(err, code){
        this.init(true); //reset state, keep auth data
        
        if(err){
            this.client.send("550 FAILED");
        }else{
            this.client.send("250 2.0.0 Ok: queued as "+(code || "FOOBARBAZ"));
        }
        
    }).bind(this));
};
